plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueTableOperationsInsertFailure.java

```diff
@@ -0,0 +1,145 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.glue;
+
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
+import com.amazonaws.services.glue.AWSGlueAsync;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import io.trino.Session;
+import io.trino.plugin.hive.metastore.Database;
+import io.trino.plugin.hive.metastore.glue.DefaultGlueColumnStatisticsProviderFactory;
+import io.trino.plugin.hive.metastore.glue.GlueHiveMetastore;
+import io.trino.plugin.hive.metastore.glue.GlueHiveMetastoreConfig;
+import io.trino.plugin.iceberg.TestingIcebergConnectorFactory;
+import io.trino.spi.security.PrincipalType;
+import io.trino.testing.AbstractTestQueryFramework;
+import io.trino.testing.LocalQueryRunner;
+import org.apache.iceberg.exceptions.CommitStateUnknownException;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Proxy;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static com.google.inject.util.Modules.EMPTY_MODULE;
+import static io.trino.plugin.hive.HiveTestUtils.HDFS_ENVIRONMENT;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static io.trino.testing.sql.TestTable.randomTableSuffix;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/*
+ * The test currently uses AWS Default Credential Provider Chain,
+ * See https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default
+ * on ways to set your AWS credentials which will be needed to run this test.
+ */
+public class TestIcebergGlueTableOperationsInsertFailure
+        extends AbstractTestQueryFramework
+{
+    private static final Logger LOG = Logger.get(TestIcebergGlueTableOperationsInsertFailure.class);
+
+    private static final String ICEBERG_CATALOG = "iceberg";
+
+    private final String schemaName = "test_iceberg_glue_" + randomTableSuffix();
+
+    private Path dataDirectory;
+
+    private GlueHiveMetastore glueHiveMetastore;
+
+    @Override
+    protected LocalQueryRunner createQueryRunner()
+            throws Exception
+    {
+        Session session = testSessionBuilder()
+                .setCatalog(ICEBERG_CATALOG)
+                .setSchema(schemaName)
+                .build();
+        LocalQueryRunner queryRunner = LocalQueryRunner.create(session);
+
+        AWSGlueAsyncAdapterProvider awsGlueAsyncAdapterProvider = delegate -> (AWSGlueAsync) Proxy.newProxyInstance(
```

Comment:
FYI

A slightly similar approach which would be more "compiler" friendly would be to use Mockito spy functionality.
https://www.javadoc.io/doc/org.mockito/mockito-core/3.0.0/org/mockito/Mockito.html#spy

However I haven't found any usage in the trino codebase of this method so I'm rather reluctant whether this approach has any chances to pass the review.